pkg/avx/collector.go

```diff
@@ -70,118 +74,229 @@ var descriptors = [numDescriptors]*prometheus.Desc{
+
+		asm.FnGetCurrentCgroupId.Call(),
+		asm.StoreMem(asm.RFP, -16, asm.R0, asm.DWord),
+		asm.Mov.Reg(asm.R2, asm.RFP),
+		asm.Add.Imm(asm.R2, -16),
+
+		// did timestamp change?
+		asm.LoadMapPtr(asm.R1, collectionMaps["avx_timestamp"].FD()),
+		asm.FnMapLookupElem.Call(),
+		asm.Mov.Imm(asm.R1, 0),
+		asm.JEq.Imm(asm.R0, 0, "timestamp"),
+		asm.LoadMem(asm.R1, asm.R0, 0, asm.Word),
+		asm.LSh.Imm(asm.R1, 32).Sym("timestamp"),
+		asm.RSh.Imm(asm.R1, 32),
+		asm.LoadMem(asm.R2, asm.RFP, -4, asm.Word),
+		asm.JEq.Reg(asm.R2, asm.R1, "exit"),
+
+		// update timestamp
+		asm.Mov.Reg(asm.R2, asm.RFP),
+		asm.Add.Imm(asm.R2, -16),
+		asm.Mov.Reg(asm.R3, asm.RFP),
+		asm.Add.Imm(asm.R3, -4),
+		asm.LoadMapPtr(asm.R1, collectionMaps["avx_timestamp"].FD()),
+		asm.Mov.Imm(asm.R4, 0),
+		asm.FnMapUpdateElem.Call(),
+
+		// last CPU
+		asm.LoadMem(asm.R3, asm.R6, 8, asm.DWord),
+		asm.Mov.Reg(asm.R6, asm.RFP),
+		asm.Add.Imm(asm.R6, -20),
+		asm.Mov.Reg(asm.R1, asm.R6),
+		asm.Mov.Imm(asm.R2, 4),
+		asm.FnProbeRead.Call(),
+
+		asm.Mov.Imm(asm.R7, 1),
+		asm.StoreMem(asm.RFP, -24, asm.R7, asm.Word),
+		asm.LoadMapPtr(asm.R1, collectionMaps["cpu_hash"].FD()),
+		asm.Mov.Reg(asm.R2, asm.R6),
+		asm.FnMapLookupElem.Call(),
+
+		// update counter atomically
+		asm.JEq.Imm(asm.R0, 0, "skip-atomic1"),
+		asm.StoreXAdd(asm.R0, asm.R7, asm.Word),
+		asm.Ja.Label("next-map"),
+
+		asm.Mov.Reg(asm.R2, asm.RFP).Sym("skip-atomic1"),
+		asm.Add.Imm(asm.R2, -20),
+		asm.Mov.Reg(asm.R3, asm.RFP),
+		asm.Add.Imm(asm.R3, -24),
+		asm.LoadMapPtr(asm.R1, collectionMaps["cpu_hash"].FD()),
+		asm.Mov.Imm(asm.R4, 0),
+		asm.FnMapUpdateElem.Call(),
+
+		asm.Mov.Reg(asm.R2, asm.RFP).Sym("next-map"),
+		asm.Add.Imm(asm.R2, -16),
+		asm.LoadMapPtr(asm.R1, collectionMaps["avx_context_switch_count"].FD()),
+		asm.FnMapLookupElem.Call(),
+
+		// update counter atomically
+		asm.JEq.Imm(asm.R0, 0, "skip-atomic2"),
+		asm.Mov.Imm(asm.R1, 1),
+		asm.StoreXAdd(asm.R0, asm.R1, asm.Word),
+		asm.Ja.Label("get-ns"),
+
+		asm.Mov.Reg(asm.R2, asm.RFP).Sym("skip-atomic2"),
+		asm.Add.Imm(asm.R2, -16),
+		asm.Mov.Reg(asm.R3, asm.RFP),
+		asm.Add.Imm(asm.R3, -24),
+		asm.LoadMapPtr(asm.R1, collectionMaps["avx_context_switch_count"].FD()),
+		asm.Mov.Imm(asm.R4, 0),
+		asm.FnMapUpdateElem.Call(),
+
+		asm.FnKtimeGetNs.Call().Sym("get-ns"),
+		asm.StoreMem(asm.RFP, -32, asm.R0, asm.DWord),
+		asm.Mov.Reg(asm.R2, asm.RFP),
+		asm.Add.Imm(asm.R2, -16),
+		asm.Mov.Reg(asm.R3, asm.RFP),
+		asm.Add.Imm(asm.R3, -32),
+		asm.LoadMapPtr(asm.R1, collectionMaps["last_update_ns"].FD()),
+		asm.Mov.Imm(asm.R4, 0),
+		asm.FnMapUpdateElem.Call(),
+
+		// return value and exit
+		asm.Mov.Imm(asm.R0, 0).Sym("exit"),
+		asm.Return(),
 	}
 
-	lastCPUCounters := bpfModule.Map("cpu")
-	if lastCPUCounters == nil {
-		return nil, errors.New("map cpu not found")
+	prog, err := bpf.NewProgram(&bpf.ProgramSpec{
+		Name:         "avx_collector",
+		Type:         bpf.TracePoint,
+		License:      "GPL",
```

Comment:
Was the original code GPL? You should indicate the licensing in the file header / `COPYING`...

Reply:
I wanted to have Apache here but that triggers:
`cannot call GPL-restricted function from non-GPL compatible program` from the kernel validator.